File: apps/releaf/android/app/src/main/java/app/releaf/mobile/data/panchanga/PanchangaRepository.kt
/*
 * PanchangaRepository.kt
 *
 * Owns the lifecycle of the bundled Vontikoppal / Mysore Panchanga
 * dataset: parses `assets/panchanga_2026_27.csv` on first launch,
 * caches every row in Room (`panchanga` table), and exposes
 * date-keyed / month-keyed / search Flows the UI consumes.
 *
 * Bootstrap is fire-and-forget on first call to `ensureLoaded()` —
 * subsequent calls short-circuit on the row count from the DAO so
 * we don't re-parse on every screen open. The CSV is small
 * (≈400 rows, ≈17KB) so parsing happens entirely in memory off the
 * main dispatcher.
 *
 * Intentionally NOT in scope here: network refresh from the GitHub
 * raw URL. The dataset is annual, the file is bundled in the APK,
 * and a network round-trip would just add a loading state to a
 * surface that's expected to work offline. Add `refreshFromNetwork`
 * later if the user needs a newer year mid-cycle.
 */

package app.releaf.mobile.data.panchanga

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader

class PanchangaRepository(
    private val context: Context,
    private val dao: PanchangaDao,
) {

    /**
     * Bootstrap the DB from the bundled asset if needed.
     *
     * Two triggers reload the table:
     *   1. The DB is empty (fresh install or wiped data).
     *   2. The bundled asset has a higher [ASSET_VERSION] than the
     *      one we last imported — used to push CSV patches to
     *      existing installs (e.g. filling in a festival the OCR
     *      pass missed) without requiring uninstall + reinstall.
     *
     * Bumping ASSET_VERSION whenever the CSV changes is the only
     * maintenance needed; the rest of the flow handles itself.
     * Errors are swallowed and logged; the UI's "data not
     * available" fallback covers the failure case.
     */
    suspend fun ensureLoaded() {
        val storedVersion = prefs.getInt(KEY_ASSET_VERSION, 0)
        val needsReload = dao.count() == 0 || storedVersion < ASSET_VERSION
        if (!needsReload) return
        withContext(Dispatchers.IO) {
            try {
                val rows = parseCsvFromAsset(ASSET_FILENAME)
                if (rows.isNotEmpty()) {
                    // `deleteAll` first so a row that disappeared
                    // between asset versions (e.g. an OCR-derived
                    // entry that was wrong and got removed) leaves
                    // the table on the next bootstrap. `upsertAll`
                    // alone wouldn't catch that case.
                    dao.deleteAll()
                    dao.upsertAll(rows)
                    prefs.edit().putInt(KEY_ASSET_VERSION, ASSET_VERSION).apply()
                }
            } catch (e: Exception) {
                // Non-fatal — the screen renders the
                // "Panchanga data not available" placeholder.
                android.util.Log.w(TAG, "Failed to bootstrap panchanga from asset", e)
            }
        }
    }

    private val prefs by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    fun observeForDate(date: String): Flow<List<PanchangaEntity>> =
        dao.observeForDate(date)

    /**
     * `monthPrefix` must be `yyyy-MM-` (with the dash) for the LIKE
     * to scope to one month exactly.
     */
    fun observeForMonth(monthPrefix: String): Flow<List<PanchangaEntity>> =
        dao.observeForMonth(monthPrefix)

    /**
     * Free-form festival search. Splits the query into whitespace-
     * separated tokens and AND-matches each one against a per-row
     * "haystack" composed of the festival name + the panchanga
     * cells that name it (special_day + masa + paksha + thithi).
     *
     * The multi-column haystack is what makes "krishna janmashtami"
     * resolve to the Bhadrapada Krishna Ashtami row: "krishna" hits
     * the `paksha` cell while "janmashtami" hits the `special_day`
     * cell. A single-LIKE SQL query couldn't span columns without
     * dynamic SQL, so this is filtered in Kotlin over the
     * `observeAllSpecialDays` flow.
     *
     * Matching has two layers, applied in order per token / word:
     *   1. Transliteration normalise → substring contains.
     *      Catches things like "janmastami" finding "janmashtami"
     *      (sh→s collapses both to the same form).
     *   2. Word-level Levenshtein with a length-scaled threshold.
     *      Catches plain typos — "gnadhi" finds "gandhi", "gandi"
     *      finds "gandhi" — without making short tokens dangerously
     *      promiscuous (a 3-letter token requires an exact match;
     *      4-7 letters tolerates one edit; ≥8 letters tolerates two).
     */
    fun searchSpecialDay(query: String): Flow<List<PanchangaEntity>> {
        val tokens = query.trim().lowercase()
            .split(WHITESPACE)
            .filter { it.isNotEmpty() }
            .map(::normalise)
        if (tokens.isEmpty()) return flowOf(emptyList())
        return dao.observeAllSpecialDays().map { rows ->
            rows.filter { row ->
                val haystack = buildString {
                    append(row.specialDayLowercase)
                    append(' ')
                    append(row.masa.lowercase())
                    append(' ')
                    append(row.paksha.lowercase())
                    append(' ')
                    append(row.thithi.lowercase())
                }.let(::normalise)
                val haystackWords = haystack.split(WHITESPACE).filter { it.isNotEmpty() }
                tokens.all { token -> matchesToken(haystack, haystackWords, token) }
            }
        }
    }

    /**
     * Token match against a row. Substring contains comes first
     * because it's both cheap and forgiving for partial inputs
     * (the user typing "krish" matches "krishna"). If that fails,
     * fall back to word-level Levenshtein so single-edit typos
     * still find the row.
     */
    private fun matchesToken(
        haystack: String,
        haystackWords: List<String>,
        token: String,
    ): Boolean {
        if (haystack.contains(token)) return true
        val threshold = typoThreshold(token.length)
        if (threshold == 0) return false
        return haystackWords.any { word -> levenshtein(word, token) <= threshold }
    }

    /**
     * Edit-distance budget by token length. Tight for short tokens
     * to avoid "abc" matching half the dataset; relaxed for longer
     * tokens where a single typo (missed letter, swapped letter)
     * is the realistic mistake.
     */
    private fun typoThreshold(tokenLength: Int): Int = when {
        tokenLength <= 3 -> 0
        tokenLength <= 7 -> 1
        else             -> 2
    }

    /**
     * Iterative Levenshtein distance with two rolling rows
     * (O(n) memory). Counts insertions, deletions, and
     * substitutions equally — no transposition shortcut, so
     * "form" → "from" costs 2.
     */
    private fun levenshtein(a: String, b: String): Int {
        if (a.isEmpty()) return b.length
        if (b.isEmpty()) return a.length
        // Quick reject: if the length difference already exceeds
        // any reasonable threshold, the distance does too.
        if (kotlin.math.abs(a.length - b.length) > MAX_TYPO_THRESHOLD) {
            return a.length + b.length
        }
        var prev = IntArray(b.length + 1) { it }
        var curr = IntArray(b.length + 1)
        for (i in 1..a.length) {
            curr[0] = i
            for (j in 1..b.length) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                curr[j] = minOf(
                    prev[j] + 1,         // deletion
                    curr[j - 1] + 1,     // insertion
                    prev[j - 1] + cost,  // substitution
                )
            }
            val tmp = prev
            prev = curr
            curr = tmp
        }
        return prev[b.length]
    }

    /**
     * Normalise a romanised Indic word so common transliteration
     * variants collapse to one form. Run on both the search
     * haystack and each query token so the user can spell things
     * however they're used to seeing them written.
     *
     * Rules (ordered — longer / more specific patterns first):
     *
     *   sh → s    Krishna   ↔ Krisna,    Janmashtami ↔ Janmastami
     *   chh → ch  Chhath    ↔ Chath
     *   ph → f    Phalguna  ↔ Falguna
     *   aa → a    Bhadraapada ↔ Bhadrapada
     *   ee → i    Sree      ↔ Sri
     *   oo → u    Pooja     ↔ Puja
     *   ii → i    Vasanti   ↔ Vasanti (no-op for already-i)
     *   uu → u    Guruu     ↔ Guru
     *
     * The list isn't exhaustive — variants like "Diwali" vs
     * "Deepavali" and "Krishn" vs "Krishna" need real lexical
     * knowledge to bridge, not substitutions. But the rules above
     * cover most of the day-to-day mismatches users hit when
     * typing festival names from memory.
     */
    private fun normalise(s: String): String {
        var out = s
        out = out.replace("chh", "ch")
        out = out.replace("aa", "a")
        out = out.replace("ee", "i")
        out = out.replace("oo", "u")
        out = out.replace("ii", "i")
        out = out.replace("uu", "u")
        out = out.replace("sh", "s")
        out = out.replace("ph", "f")
        return out
    }

    // ---------- CSV parsing ----------

    /**
     * Parses the bundled CSV. Handles the three shapes observed in
     * the upstream dataset:
     *   - 6 fields, last (special_day) empty:   `2026-03-21,Chaitra,Shukla,Tritiya,3,`
     *   - 5 fields, trailing comma omitted:     `2026-04-07,Chaitra,Krishna,Panchami,5`
     *   - 6 fields, special_day quoted with
     *     internal comma:                       `2026-03-26,Chaitra,Shukla,Ashtami,8,"Ashoka Ashtami, Kuruba Jayanti"`
     *
     * Implemented inline because the file is small (~400 rows) and
     * Kotlin's standard library has no CSV parser — pulling in a
     * dependency for one file isn't worth the build-time cost.
     */
    private fun parseCsvFromAsset(filename: String): List<PanchangaEntity> {
        context.assets.open(filename).use { stream ->
            BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).use { reader ->
                val header = reader.readLine() ?: return emptyList()
                require(header.startsWith("date,")) {
                    "Unexpected CSV header: $header"
                }

                val out = ArrayList<PanchangaEntity>(400)
                while (true) {
                    val line = reader.readLine() ?: break
                    if (line.isBlank()) continue
                    val fields = parseCsvLine(line)
                    if (fields.size < 5) continue   // malformed row — skip
                    val date       = fields[0].trim()
                    val masa       = fields[1].trim()
                    val paksha     = fields[2].trim()
                    val thithi     = fields[3].trim()
                    val thithiNum  = fields[4].trim()
                    val specialDay = if (fields.size >= 6) fields[5].trim() else ""
                    out += PanchangaEntity(
                        // `date#thithiNum` keeps duplicate-date rows
                        // distinguishable (e.g. 2026-07-28 has
                        // Chaturdashi and Purnima both falling on
                        // the same Gregorian day).
                        id                  = "$date#$thithiNum",
                        date                = date,
                        masa                = masa,
                        paksha              = paksha,
                        thithi              = thithi,
                        thithiNum           = thithiNum,
                        specialDay          = specialDay,
                        specialDayLowercase = specialDay.lowercase(),
                    )
                }
                return out
            }
        }
    }

    /**
     * Single-line CSV parser — handles double-quoted fields with
     * embedded commas. Doesn't handle escaped quotes inside quoted
     * fields ("") because the upstream dataset never uses them; if
     * future rows do, extend here.
     */
    private fun parseCsvLine(line: String): List<String> {
        val fields = ArrayList<String>(6)
        val sb = StringBuilder()
        var inQuotes = false
        var i = 0
        while (i < line.length) {
            val c = line[i]
            when {
                c == '"' -> inQuotes = !inQuotes
                c == ',' && !inQuotes -> {
                    fields += sb.toString()
                    sb.setLength(0)
                }
                else -> sb.append(c)
            }
            i++
        }
        fields += sb.toString()
        return fields
    }

    private companion object {
        const val ASSET_FILENAME = "panchanga_2026_27.csv"
        const val TAG = "PanchangaRepository"
        val WHITESPACE = Regex("\\s+")
        /** Highest edit-distance budget [typoThreshold] ever returns.
         *  Used by [levenshtein]'s length-difference fast-reject. */
        const val MAX_TYPO_THRESHOLD = 2

        /**
         * Bundled-asset content version. **Bump this whenever the
         * CSV changes** so existing installs re-import on next
         * launch. The version number is monotonic — the bootstrap
         * compares stored < ASSET_VERSION rather than `!=` so a
         * downgrade (impossible in practice, but) doesn't trigger
         * a reload.
         *
         *   v1: initial OCR import.
         *   v2: patched 2026-09-04 with Sri Krishna Janmashtami,
         *       which the OCR pass missed.
         *   v3: patched six more OCR-missed festivals —
         *       Hanuman Jayanti (2026-04-02), Akshaya Tritiya
         *       (2026-04-20), Buddha Purnima (2026-05-01),
         *       Mahalaya Amavasya (2026-10-10), Diwali & Sri
         *       Lakshmi Puja (2026-11-09), Christmas (2026-12-25).
         */
        const val ASSET_VERSION = 3

        const val PREFS_NAME = "panchanga_repo"
        const val KEY_ASSET_VERSION = "asset_version"
    }
}

File: apps/quickink/ios/QuickInk/Data/Panchanga/PanchangaRepository.swift
/*
 * PanchangaRepository.swift
 *
 * Owns the lifecycle of the bundled Vontikoppal / Mysore Panchanga
 * dataset: parses `Resources/panchanga_2026_27.csv` on first launch,
 * caches every row in the `panchanga` SQLite table, and exposes
 * date-keyed / month-keyed / search publishers the UI consumes.
 *
 * Bootstrap is fire-and-forget on first call to `ensureLoaded()` —
 * subsequent calls short-circuit on the row count and a stored asset
 * version. The CSV is small (≈400 rows, 17 KB) so parsing happens
 * entirely in memory off the main actor.
 *
 * Port of Releaf Android's `PanchangaRepository.kt`. Same parser, same
 * search semantics (transliteration normalise + Levenshtein typo
 * tolerance), GRDB ValueObservation instead of Room Flow.
 */

import Foundation
import Combine
import GRDB

public final class PanchangaRepository {

    /// Bundled CSV asset version. **Bump this whenever the CSV changes**
    /// so existing installs re-import on next launch. Mirror of the
    /// Android `ASSET_VERSION` constant; current value matches v3
    /// (initial OCR + 7 festival patches).
    public static let assetVersion: Int = 3
    private static let assetFilename = "panchanga_2026_27"
    private static let assetExtension = "csv"
    private static let userDefaultsKey = "quickink.panchanga.asset_version"

    private let dbQueue: DatabaseQueue

    public init(database: QuickInkDatabase = .shared) {
        self.dbQueue = database.dbQueue
    }

    // MARK: - Bootstrap

    /// Idempotently seed the `panchanga` table from the bundled CSV
    /// when needed. Two triggers reload:
    ///   1. The table is empty (fresh install / wiped data).
    ///   2. The stored asset version is older than `assetVersion` —
    ///      lets us push CSV patches to existing installs without
    ///      requiring an uninstall.
    public func ensureLoaded() async {
        do {
            let count = try await dbQueue.read { db in
                try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM panchanga") ?? 0
            }
            let stored = UserDefaults.standard.integer(forKey: Self.userDefaultsKey)
            if count > 0 && stored >= Self.assetVersion { return }

            let rows = try parseCsvFromBundle()
            guard !rows.isEmpty else { return }

            try await dbQueue.write { db in
                try db.execute(sql: "DELETE FROM panchanga")
                for row in rows {
                    try row.insert(db)
                }
            }
            UserDefaults.standard.set(Self.assetVersion, forKey: Self.userDefaultsKey)
        } catch {
            // Non-fatal — the screen renders the "Panchanga data not
            // available" placeholder when the table stays empty.
            NSLog("[panchanga] ensureLoaded failed: %@", "\(error)")
        }
    }

    // MARK: - Observation

    /// Rows for the given Gregorian date, ordered by tithi number.
    /// Some dates carry two rows (transition days).
    public func observeForDate(_ date: String) -> AnyPublisher<[PanchangaEntity], Never> {
        ValueObservation.tracking { db in
            try PanchangaEntity.fetchAll(db, sql: """
                SELECT * FROM panchanga
                WHERE date = ?
                ORDER BY thithi_num ASC
                """, arguments: [date])
        }
        .publisher(in: dbQueue)
        .catch { _ in Just<[PanchangaEntity]>([]) }
        .eraseToAnyPublisher()
    }

    /// Rows whose `date` starts with `monthPrefix` (e.g. "2026-05-").
    /// Powers the festival-dot indicators on the calendar grid.
    public func observeForMonth(prefix monthPrefix: String) -> AnyPublisher<[PanchangaEntity], Never> {
        ValueObservation.tracking { db in
            try PanchangaEntity.fetchAll(db, sql: """
                SELECT * FROM panchanga
                WHERE date LIKE ? || '%'
                ORDER BY date ASC, thithi_num ASC
                """, arguments: [monthPrefix])
        }
        .publisher(in: dbQueue)
        .catch { _ in Just<[PanchangaEntity]>([]) }
        .eraseToAnyPublisher()
    }

    /// All rows that carry a non-empty `special_day`. The search
    /// implementation filters in-process so we can AND-match tokens
    /// across `special_day + masa + paksha + thithi`.
    public func observeAllSpecialDays() -> AnyPublisher<[PanchangaEntity], Never> {
        ValueObservation.tracking { db in
            try PanchangaEntity.fetchAll(db, sql: """
                SELECT * FROM panchanga
                WHERE special_day_lc != ''
                ORDER BY date ASC, thithi_num ASC
                """)
        }
        .publisher(in: dbQueue)
        .catch { _ in Just<[PanchangaEntity]>([]) }
        .eraseToAnyPublisher()
    }

    // MARK: - Festival search

    /// Free-form festival search. Splits the query into whitespace-
    /// separated tokens and AND-matches each one against a per-row
    /// "haystack" composed of `special_day + masa + paksha + thithi`.
    ///
    /// Matching has two layers per token:
    ///   1. Transliteration-normalised substring (e.g. "janmastami"
    ///      finds "janmashtami" via `sh→s`).
    ///   2. Word-level Levenshtein with a length-scaled threshold
    ///      (≤3 letters: exact; 4-7: one edit; ≥8: two edits).
    public func searchSpecialDay(_ query: String) -> AnyPublisher<[PanchangaEntity], Never> {
        let tokens = query
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
            .filter { !$0.isEmpty }
            .map(Self.normalise)
        if tokens.isEmpty {
            return Just<[PanchangaEntity]>([]).eraseToAnyPublisher()
        }
        return observeAllSpecialDays()
            .map { rows -> [PanchangaEntity] in
                rows.filter { row in
                    let haystackRaw = "\(row.specialDayLowercase) \(row.masa.lowercased()) \(row.paksha.lowercased()) \(row.thithi.lowercased())"
                    let haystack = Self.normalise(haystackRaw)
                    let haystackWords = haystack
                        .split(whereSeparator: { $0.isWhitespace })
                        .map(String.init)
                        .filter { !$0.isEmpty }
                    return tokens.allSatisfy { token in
                        Self.matchesToken(haystack: haystack, haystackWords: haystackWords, token: token)
                    }
                }
            }
            .eraseToAnyPublisher()
    }

    private static func matchesToken(haystack: String, haystackWords: [String], token: String) -> Bool {
        if haystack.contains(token) { return true }
        let threshold = typoThreshold(token.count)
        if threshold == 0 { return false }
        return haystackWords.contains { levenshtein($0, token) <= threshold }
    }

    private static func typoThreshold(_ tokenLength: Int) -> Int {
        switch tokenLength {
        case 0...3: return 0
        case 4...7: return 1
        default:    return 2
        }
    }

    private static let maxTypoThreshold = 2

    private static func levenshtein(_ a: String, _ b: String) -> Int {
        let ac = Array(a)
        let bc = Array(b)
        if ac.isEmpty { return bc.count }
        if bc.isEmpty { return ac.count }
        if abs(ac.count - bc.count) > maxTypoThreshold {
            return ac.count + bc.count
        }
        var prev = Array(0...bc.count)
        var curr = Array(repeating: 0, count: bc.count + 1)
        for i in 1...ac.count {
            curr[0] = i
            for j in 1...bc.count {
                let cost = ac[i - 1] == bc[j - 1] ? 0 : 1
                curr[j] = Swift.min(
                    prev[j] + 1,         // deletion
                    curr[j - 1] + 1,     // insertion
                    prev[j - 1] + cost   // substitution
                )
            }
            swap(&prev, &curr)
        }
        return prev[bc.count]
    }

    /// Collapse common romanised-Indic transliteration variants —
    /// `sh→s`, `chh→ch`, `ph→f`, `aa→a`, `ee→i`, etc. — so the user
    /// can spell festival names however they're used to seeing them.
    /// Ordered longest / most specific first.
    private static func normalise(_ s: String) -> String {
        var out = s
        out = out.replacingOccurrences(of: "chh", with: "ch")
        out = out.replacingOccurrences(of: "aa",  with: "a")
        out = out.replacingOccurrences(of: "ee",  with: "i")
        out = out.replacingOccurrences(of: "oo",  with: "u")
        out = out.replacingOccurrences(of: "ii",  with: "i")
        out = out.replacingOccurrences(of: "uu",  with: "u")
        out = out.replacingOccurrences(of: "sh",  with: "s")
        out = out.replacingOccurrences(of: "ph",  with: "f")
        return out
    }

    // MARK: - CSV parsing

    private func parseCsvFromBundle() throws -> [PanchangaEntity] {
        guard let url = Bundle.module.url(forResource: Self.assetFilename, withExtension: Self.assetExtension) else {
            return []
        }
        let raw = try String(contentsOf: url, encoding: .utf8)
        var lines = raw.split(whereSeparator: \.isNewline).map(String.init)
        guard !lines.isEmpty else { return [] }
        let header = lines.removeFirst()
        guard header.hasPrefix("date,") else { return [] }

        var out: [PanchangaEntity] = []
        out.reserveCapacity(400)
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty { continue }
            let fields = Self.parseCsvLine(trimmed)
            if fields.count < 5 { continue }
            let date       = fields[0].trimmingCharacters(in: .whitespaces)
            let masa       = fields[1].trimmingCharacters(in: .whitespaces)
            let paksha     = fields[2].trimmingCharacters(in: .whitespaces)
            let thithi     = fields[3].trimmingCharacters(in: .whitespaces)
            let thithiNum  = fields[4].trimmingCharacters(in: .whitespaces)
            let specialDay = fields.count >= 6 ? fields[5].trimmingCharacters(in: .whitespaces) : ""
            out.append(PanchangaEntity(
                id: "\(date)#\(thithiNum)",
                date: date,
                masa: masa,
                paksha: paksha,
                thithi: thithi,
                thithiNum: thithiNum,
                specialDay: specialDay,
                specialDayLowercase: specialDay.lowercased()
            ))
        }
        return out
    }

    /// Single-line CSV parser — handles double-quoted fields with
    /// embedded commas. Doesn't handle escaped quotes inside quoted
    /// fields (the upstream dataset never uses them).
    private static func parseCsvLine(_ line: String) -> [String] {
        var fields: [String] = []
        fields.reserveCapacity(6)
        var current = ""
        var inQuotes = false
        for ch in line {
            if ch == "\"" {
                inQuotes.toggle()
            } else if ch == "," && !inQuotes {
                fields.append(current)
                current.removeAll(keepingCapacity: true)
            } else {
                current.append(ch)
            }
        }
        fields.append(current)
        return fields
    }
}
